### ISSUE TITLE
Add Frame option to Timer node

### ISSUE THREAD

**Issue by CoreTaxxe** [godotengine/godot-proposals#10879]
### Describe the project you are working on

Creating a doomlike, retro-styled or fighting game where frame accuracy is crucial. (For example for combos, invincibilities) where time-based timers are not sufficient or can't guarantee frame-accurate timings. A lot of new comers starting to program re-create games/programs they know. Providing them with the tools they expect seems the best way to get them started.

I am aware that you typically don't want to use frames anymore. But in some rare occurrences you might and given this feature is opt-in/optional users who want to use it can. (Especially since they could also just go through the tiny but tedious process of writing it themselves anyways)

First draft implementation:
[#97725](https://github.com/godotengine/godot/pull/97725)

### Describe the problem or limitation you are having in your project

QOL Feature to ease the use of frame timers. Otherwise, users have to create custom nodes or re-implement frame counters every single time over again. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Frame timers, different from time based timers, are directly bound to either physics or process frame rate thus guaranteeing deterministic and accurate results even through frame lags. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add an opion to the timer node that allows to chose between Frame and Time mode. In frame mode frames are counted instead of seconds.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes.

### Is there a reason why this should be core and not an add-on in the asset library?

Its a very simple to implement, has pretty much 0 need for maintenance and is too simple too be worth an addon.

**Comment by AThousandShips** [comment]
Beyond the issues with relying on fixed frame counts, this is so enormously trivial to accomplish, you just add a counter variable to your class and count it in `_process` or `_physics_process`, this avoids adding a whole separate node and using signals, much simpler and easier

**Comment by CoreTaxxe** [comment]
I agree for very simple counters that is indeed overkill. However as soon as you have more than 1 counter it gets really messy. Especially if you want pause, restart or change behavior you have to spend significant amount of code.

**Comment by AThousandShips** [comment]
I really do not expect any user to be needing that, this is an extreme edge case to begin with, and not recommended practice

**Comment by CoreTaxxe** [comment]
> I really do not expect any user to be needing that, this is an extreme edge case to begin with, and not recommended practice 

I totally agree that its not good practice for longer timers and probably useless for them. But godot engine itself recommend it for smaller counter so i really can't give you that point. (https://github.com/godotengine/godot/blob/master/scene%2Fmain%2Ftimer.cpp#L203) Frame counting is especially useful and still go-to for order validation (for example attacks)

**Comment by AThousandShips** [comment]
See [best practices for engine contributors](https://docs.godotengine.org/en/latest/contributing/development/best_practices_for_engine_contributors.html) for details on our policies and practices

**Comment by AThousandShips** [comment]
It's the answer to the question of why we don't implement things that are niche and goes against general recommendations, the engine doesn't recommend using a timer in this way, but to use other means *than* a timer

The basic rule of implementing certain features is if they are used often enough, needed often enough, and can't be trivially worked around. Even if it's minimal any additional maintenance burden is still a burden, and niche features are even worse as few might use it, making errors in it harder to notice, making things harder to maintain. That's the general practice

**Comment by CoreTaxxe** [comment]
> It's the answer to the question of why we don't implement things that are niche and goes against general recommendations, the engine doesn't recommend using a timer in this way, but to use other means *than* a timer
> 
> The basic rule of implementing certain features is if they are used often enough, needed often enough, and can't be trivially worked around. Even if it's minimal any additional maintenance burden is still a burden, and niche features are even worse as few might use it, making errors in it harder to notice, making things harder to maintain. That's the general practice

But thats not really applicable, no? You cant implement tick based systems reliably with delta time(rs). And those are still pretry common. Also having to introduce multiple variables, statements and function instead of just calling methods of an timer is actually bad practice.

**Comment by Calinou** [comment]
> I totally agree that its not good practice for longer timers and probably useless for them. But godot engine itself recommend it for smaller counter so i really can't give you that point. ([godotengine/godot@master/scene%2Fmain%2Ftimer.cpp#L203](https://github.com/godotengine/godot/blob/master/scene%2Fmain%2Ftimer.cpp?rgh-link-date=2025-01-31T12%3A37%3A11.000Z#L203)) Frame counting is especially useful and still go-to for order validation (for example attacks)

This addressed by https://github.com/godotengine/godot/pull/89786#.

**Comment by CoreTaxxe** [comment]
> > I totally agree that its not good practice for longer timers and probably useless for them. But godot engine itself recommend it for smaller counter so i really can't give you that point. ([godotengine/godot@master/scene%2Fmain%2Ftimer.cpp#L203](https://github.com/godotengine/godot/blob/master/scene%2Fmain%2Ftimer.cpp?rgh-link-date=2025-01-31T12%3A37%3A11.000Z#L203)) Frame counting is especially useful and still go-to for order validation (for example attacks)
> 
> This addressed by https://github.com/godotengine/godot/pull/89786#.


Isnt that just a tick based timer option but with rounding errors instead of frame accurate ones? Good addition nevertheless. Tho makes me wonder even more why there cant be a frame mode to have deterministics. For example if i want to invoke twice ps.

**Comment by CoreTaxxe** [comment]
> > > I totally agree that its not good practice for longer timers and probably useless for them. But godot engine itself recommend it for smaller counter so i really can't give you that point. ([godotengine/godot@master/scene%2Fmain%2Ftimer.cpp#L203](https://github.com/godotengine/godot/blob/master/scene%2Fmain%2Ftimer.cpp?rgh-link-date=2025-01-31T12%3A37%3A11.000Z#L203)) Frame counting is especially useful and still go-to for order validation (for example attacks)
> > 
> > This addressed by https://github.com/godotengine/godot/pull/89786#.
> 
> 
> Isnt that just a tick based timer option but with rounding errors instead of frame accurate ones? Good addition nevertheless. Tho makes me wonder even more why there cant be a frame mode to have deterministics. For example if i want to invoke twice ps.
> 
 Also wouldn that fall under the "inventing a problem for a solution" as well?

**Comment by Calinou** [comment]
> Tho makes me wonder even more why there cant be a frame mode to have deterministics. For example if i want to invoke twice ps.

If you need deterministic physics, you'll need to write your own physics using only fixed-point coordinates (i.e. integers, no floats). At this point, writing your own timer is only a small piece of the puzzle 🙂

**Comment by CoreTaxxe** [comment]
> > Tho makes me wonder even more why there cant be a frame mode to have deterministics. For example if i want to invoke twice ps.
> 
> If you need deterministic physics, you'll need to write your own physics using only fixed-point coordinates (i.e. integers, no floats). At this point, writing your own timer is only a small piece of the puzzle 🙂 
> 
> > Also wouldn that fall under the "inventing a problem for a solution" as well?
> 
> There are many real world cases that https://github.com/godotengine/godot/pull/89786 would already resolve, such as weapons with high fire rates firing slower at low FPS.

But having your weapon fire a few ms earlier due to a rounding error oppsoed to tick perfect comparision would void that. And if you dont need vompetet8ve accuracy you could work around with a script just as well - as ive been told originally..

**Comment by CoreTaxxe** [comment]
You know what. I give up.
